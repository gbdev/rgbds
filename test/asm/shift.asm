macro test
	; Test the rpn system, as well as the linker...
	DEF expr EQUS STRRPL(STRRPL("\1 + zero)", "<< ", "<< ("), ">> ", ">> (")
	dl expr
	PURGE expr

	; ...as well as the constexpr system
	println "\1 = ", (\1)
endm

section "test", ROM0[0]

	test 1 << 1
	test 1 << 32
	test 1 << 9001
	test -1 << 1
	test -1 << 32
	test -1 << -9001

	test -1 >> 1
	test -1 >> 32
	test -1 >> 9001
	test -4 >> 1
	test -4 >> 2
	test -1 >> -9001

	test 100 << -2
	test 1 >> -2

	test 100 >>> 16
	test 100 >>> -16
	test 100 >>> 32
	test 100 >>> -32
	test -100 >>> 32
	test -100 >>> -32

	test $DEADBEEF >> 1
	test $DEADBEEF >>> 1

SECTION "Zero", ROM0[0]
zero:
