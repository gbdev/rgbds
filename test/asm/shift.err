warning: shift.asm(14) -> shift.asm::test(8): [-Wshift-amount]
    Shifting left by large amount 32
warning: shift.asm(15) -> shift.asm::test(8): [-Wshift-amount]
    Shifting left by large amount 9001
warning: shift.asm(17) -> shift.asm::test(8): [-Wshift-amount]
    Shifting left by large amount 32
warning: shift.asm(18) -> shift.asm::test(8): [-Wshift-amount]
    Shifting left by negative amount -9001
warning: shift.asm(20) -> shift.asm::test(8): [-Wshift]
    Shifting right negative value -1
warning: shift.asm(21) -> shift.asm::test(8): [-Wshift]
    Shifting right negative value -1
warning: shift.asm(21) -> shift.asm::test(8): [-Wshift-amount]
    Shifting right by large amount 32
warning: shift.asm(22) -> shift.asm::test(8): [-Wshift]
    Shifting right negative value -1
warning: shift.asm(22) -> shift.asm::test(8): [-Wshift-amount]
    Shifting right by large amount 9001
warning: shift.asm(23) -> shift.asm::test(8): [-Wshift]
    Shifting right negative value -4
warning: shift.asm(24) -> shift.asm::test(8): [-Wshift]
    Shifting right negative value -4
warning: shift.asm(25) -> shift.asm::test(8): [-Wshift]
    Shifting right negative value -1
warning: shift.asm(25) -> shift.asm::test(8): [-Wshift-amount]
    Shifting right by negative amount -9001
warning: shift.asm(27) -> shift.asm::test(8): [-Wshift-amount]
    Shifting left by negative amount -2
warning: shift.asm(28) -> shift.asm::test(8): [-Wshift-amount]
    Shifting right by negative amount -2
warning: shift.asm(31) -> shift.asm::test(8): [-Wshift-amount]
    Shifting right by negative amount -16
warning: shift.asm(32) -> shift.asm::test(8): [-Wshift-amount]
    Shifting right by large amount 32
warning: shift.asm(33) -> shift.asm::test(8): [-Wshift-amount]
    Shifting right by negative amount -32
warning: shift.asm(34) -> shift.asm::test(8): [-Wshift-amount]
    Shifting right by large amount 32
warning: shift.asm(35) -> shift.asm::test(8): [-Wshift-amount]
    Shifting right by negative amount -32
warning: shift.asm(37) -> shift.asm::test(8): [-Wshift]
    Shifting right negative value -559038737
