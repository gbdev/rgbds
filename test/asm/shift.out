1 << 1 = $2
1 << 32 = $0
1 << 9001 = $0
-1 << 1 = $FFFFFFFE
-1 << 32 = $0
-1 << -9001 = $FFFFFFFF
-1 >> 1 = $FFFFFFFF
-1 >> 32 = $FFFFFFFF
-1 >> 9001 = $FFFFFFFF
-4 >> 1 = $FFFFFFFE
-4 >> 2 = $FFFFFFFF
-1 >> -9001 = $0
100 << -2 = $19
1 >> -2 = $4
100 >>> 16 = $0
100 >>> -16 = $640000
100 >>> 32 = $0
100 >>> -32 = $0
-100 >>> 32 = $0
-100 >>> -32 = $0
$DEADBEEF >> 1 = $EF56DF77
$DEADBEEF >>> 1 = $6F56DF77
